drivers/input/input_pinnacle.c

```diff
@@ -0,0 +1,627 @@
+	sample->x = ((rx[0] & BIT(4)) == BIT(4)) ? -(256 - rx[1]) : rx[1];
+	sample->y = ((rx[0] & BIT(5)) == BIT(5)) ? -(256 - rx[2]) : rx[2];
+#endif
+}
+
+#if defined(CONFIG_INPUT_PINNACLE_DATA_MODE_ABSOLUTE)
+static bool pinnacle_is_idle_sample(const struct pinnacle_sample *sample)
+{
+	return (sample->x == 0 && sample->y == 0 && sample->z == 0);
+}
+
+static void pinnacle_clip_sample(const struct pinnacle_active_range *range,
+				 struct pinnacle_sample *sample)
+{
+	if (sample->x < range->x_min) {
+		sample->x = range->x_min;
+	}
+	if (sample->x > range->x_max) {
+		sample->x = range->x_max;
+	}
+	if (sample->y < range->y_min) {
+		sample->y = range->y_min;
+	}
+	if (sample->y > range->y_max) {
+		sample->y = range->y_max;
+	}
+}
+
+static void pinnacle_scale_sample(const struct pinnacle_active_range *range,
+				  const struct pinnacle_resolution *res,
+				  struct pinnacle_sample *sample)
+{
+	sample->x = (uint16_t)((uint32_t)(sample->x - range->x_min) * res->x /
+			       (range->x_max - range->x_min));
+	sample->y = (uint16_t)((uint32_t)(sample->y - range->y_min) * res->y /
+			       (range->y_max - range->y_min));
+}
+#endif
+
+static int pinnacle_sample_fetch(const struct device *dev, struct pinnacle_sample *sample)
+{
+	const struct pinnacle_config *config = dev->config;
+	const struct pinnacle_bus *bus = &config->bus;
+
+	int ret1 = 0;
+	int ret2 = 0;
+
+#if defined(CONFIG_INPUT_PINNACLE_DATA_MODE_ABSOLUTE)
+	uint8_t rx[4];
+
+	ret1 = bus->seq_read(bus, PINNACLE_REG_PACKET_BYTE2, rx, 4);
+#elif defined(CONFIG_INPUT_PINNACLE_DATA_MODE_RELATIVE)
+	uint8_t rx[3];
+
+	ret1 = bus->seq_read(bus, PINNACLE_REG_PACKET_BYTE0, rx, 3);
+#endif
+
+	if (ret1) {
+		LOG_ERR("Failed to read data from SPI device");
+	} else {
+		pinnacle_decode_sample(rx, sample);
+	}
+	ret2 = bus->write(bus, PINNACLE_REG_STATUS1, 0x00);
+	if (ret2) {
+		LOG_ERR("Failed to clear SW_CC and SW_DR for SPI device");
+		return ret2;
+	}
+	return 0;
+}
+
+static int pinnacle_handle_interrupt(const struct device *dev)
+{
+	struct pinnacle_data *drv_data = dev->data;
+	struct pinnacle_sample *sample = &drv_data->sample;
+
+	int ret = pinnacle_sample_fetch(dev, sample);
+
+	if (ret) {
+		LOG_ERR("Failed to read data packets");
+	}
+
+#if defined(CONFIG_INPUT_PINNACLE_DATA_MODE_ABSOLUTE)
+	const struct pinnacle_config *config = dev->config;
+
+	if (config->clipping_enabled && !pinnacle_is_idle_sample(sample)) {
+		pinnacle_clip_sample(&config->active_range, sample);
+		if (config->scaling_enabled) {
+			pinnacle_scale_sample(&config->active_range, &config->resolution, sample);
+		}
+	}
+
+	input_report_abs(dev, INPUT_ABS_X, sample->x, false, K_FOREVER);
+	input_report_abs(dev, INPUT_ABS_Y, sample->y, false, K_FOREVER);
+	input_report_abs(dev, INPUT_ABS_Z, sample->z, true, K_FOREVER);
+#elif defined(CONFIG_INPUT_PINNACLE_DATA_MODE_RELATIVE)
+	input_report_abs(dev, INPUT_REL_X, sample->x, false, K_FOREVER);
+	input_report_abs(dev, INPUT_REL_Y, sample->y, true, K_FOREVER);
```

Comment:
how about making these three configurable? then one could swap x/y by changing the codes

Reply:
sorry I'm asking about making the code configurable, like instead of hardcoding INPUT_REL_X/INPUT_REL_Y having a code passed from the devicetree so one could swap the axis, like https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/input/pixart%2Cpmw3610.yaml#L22-L34.

Also just realized, the rel one should use `input_report_rel`

Reply:
I addressed the issue with reporting relative coordinates, `input_report_rel` is used for them now. I'll play with making the reporting mode configurable and ome back with results.